opencog/query/Distributed/DistSCM.cc

```diff
@@ -0,0 +1,230 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#define MAX_SCM_CODE_STRING 5000
+
+#include "DistSCM.h"
+#include <opencog/guile/SchemeEval.h>//which lib to link?
+#include <opencog/guile/SchemeSmob.h>
+
+using namespace opencog;
+
+bool DistSCM::master_mode=true;
+
+DistSCM::DistSCM(void):true_string("true"),false_string("false")
+{
+  scm_with_guile(init_in_guile, this);
+}
+
+void* DistSCM::init_in_guile(void* self)
+{
+    scm_c_define_module("opencog dist-gearman", init_in_module, self);
+    scm_c_use_module("opencog dist-gearman");
+    return NULL;
+}
+
+void DistSCM::init_in_module(void* data)
+{
+    DistSCM* self = (DistSCM*) data;
+    self->init();
+}
+
+void DistSCM::init(void)
+{
+  //blocking slave mode
+	define_scheme_primitive("set-slave-mode", &DistSCM::slave_mode,
+                            this, "dist-gearman");
+  //exits the slave mode after last request is served
+	define_scheme_primitive("set-master-mode", &DistSCM::set_master_mode,
+                            this, "dist-gearman");
+  //sends scheme string to slave and blocks till result arrives, returns uuid of resulting handle, invalid_uuid on failure
+  define_scheme_primitive("dist-run-scm", &DistSCM::dist_scm,
+                            this, "dist-gearman");
+}
+
+void DistSCM::set_master_mode(void)
+{
+	master_mode=true;
+}
+
+gearman_return_t DistSCM::worker_function(gearman_job_st *job, void *context)
+{
+  const char *workload= (const char *)gearman_job_workload(job);
+  const size_t workload_size= gearman_job_workload_size(job);
+
+  //std::cout << "Recieved " << workload_size << " bytes" << std::endl;
+  if (workload_size>MAX_SCM_CODE_STRING+1) return GEARMAN_ERROR;
+  char scm_str[MAX_SCM_CODE_STRING+1];
+  strcpy(scm_str,workload);
+
+  SchemeEval evl((AtomSpace*)context);
+  evl.init_scheme();
+  //std::cout<<scm_str<<"\n";
+  Handle result=evl.eval_h(scm_str);//can we check if handle is valid? how to get current atomspace
+  UUID vl=result.value();
+  if (Handle::UNDEFINED==result)std::cerr<<"invalid handle \n";
+
+  evl.eval("(sql-store)");
+
+  if (gearman_failed(gearman_job_send_data(job, &vl, sizeof(vl))))
+  {
+		std::cerr<<"error occured";
+    return GEARMAN_ERROR;//on gearman error return, no result is sent to master and job request stays in que
+  }
+   //std::cout<<"gearman data sent..."<<"\n";
+   return GEARMAN_SUCCESS;
+}
+
+const std::string& DistSCM::slave_mode(const std::string& ip_string,const std::string& workerID)
+{
+	AtomSpace* atoms_ptr=SchemeSmob::ss_get_env_as("set-slave-mode");
+	//std::cout<<"atomspace "<<atoms_ptr<<"\n";
+
+    if ((worker= gearman_worker_create(NULL)) == NULL)
+    {
+      std::cerr << "Memory allocation failure on worker creation." << std::endl;
+      return false_string;
```

Comment:
Hi Linas,
Thanks for the review, I will make those changes.
Regards,
Mandeep
On Sep 24, 2015 1:29 AM, "Linas Vepštas" <notifications@github.com> wrote:

> In opencog/query/Distributed/DistSCM.cc
> <https://github.com/opencog/atomspace/pull/328#discussion_r40232175>:
>
> > +		std::cerr<<"error occured";
> > +    return GEARMAN_ERROR;//on gearman error return, no result is sent to master and job request stays in que
> > +  }
> > +   //std::cout<<"gearman data sent..."<<"\n";
> > +   return GEARMAN_SUCCESS;
> > +}
> > +
> > +const std::string& DistSCM::slave_mode(const std::string& ip_string,const std::string& workerID)
> > +{
> > +	AtomSpace* atoms_ptr=SchemeSmob::ss_get_env_as("set-slave-mode");
> > +	//std::cout<<"atomspace "<<atoms_ptr<<"\n";
> > +
> > +    if ((worker= gearman_worker_create(NULL)) == NULL)
> > +    {
> > +      std::cerr << "Memory allocation failure on worker creation." << std::endl;
> > +      return false_string;
>
> I mean, Kurt Akeley personally showed me how crazy error checking was, in
> 1990. We had to engineer a hackfest around this, because at the time,
> compilers did not yet support exceptions. We invented something very
> clever: it was almost like an exception, but it did not need compiler
> support. Later on, Kurt got awards for this and many other inventions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/atomspace/pull/328/files#r40232175>.
>